packages/core/src/molecule/codec.advanced.ts

```diff
@@ -0,0 +1,373 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+
+import { bytesConcat, bytesFrom, BytesLike } from "../bytes/index.js";
+import { bytes, Codec, EncodableType, fixedBytes } from "./codec.js";
+import { Uint32LE } from "./predefined.js";
+
+/**
+ * Vector with fixed size item codec
+ * @param itemCodec fixed-size vector item codec
+ */
+export function fixedItemVec<Encodable>(
+  itemCodec: Codec<Encodable>,
+): Codec<Array<Encodable>> {
+  if (itemCodec.byteLength === undefined) {
+    throw new Error("fixedItemVec: itemCodec requires a byte length");
+  }
+  return bytes({
+    encode(userDefinedItems) {
+      const encodedArray = userDefinedItems.map((item) =>
+        itemCodec.encode(item),
+      );
+      return bytesConcat(
+        Uint32LE.encode(BigInt(userDefinedItems.length)),
+        encodedArray.reduce(
+          (concated, item) => bytesConcat(concated, item),
+          new ArrayBuffer(0),
+        ),
+      );
+    },
+    decode(buffer) {
+      const value = bytesFrom(buffer);
+      if (value.byteLength < 4) {
+        throw new Error(
+          `fixedItemVec: too short buffer, expected at least 4 bytes, but got ${value.byteLength}`,
+        );
+      }
+      const itemCount = Number(Uint32LE.decode(value.slice(0, 4)));
+      const itemByteLength = itemCodec.byteLength!;
+      const byteLength = 4 + itemCount * itemByteLength;
+      if (value.byteLength !== byteLength) {
+        throw new Error(
+          `fixedItemVec: invalid buffer size, expected ${byteLength}, but got ${value.byteLength}`,
+        );
+      }
+      const decodedArray: Array<Encodable> = [];
+      for (let offset = 0; offset < byteLength; offset += itemByteLength) {
+        decodedArray.push(
+          itemCodec.decode(value.slice(offset, offset + itemByteLength)),
```

Comment:
Can we catch exceptions here and modify the message? e.g. If the itemCodec throws "ItemError: Invalid data", we can wrap it to "FixedItemVec(ItemError: Invalid data)" here for more explicit error message.